### ISSUE TITLE
What are the requirements for a script passed to the compilers option?

### ISSUE THREAD

**Issue by d4goxn** [mochajs/mocha#1458]
I would like to pass a JSX preprocessor to Mocha: `mocha --compilers js:./preprocessor.js`. I haven't been able to find any documentation about writing Mocha compilers, so I just tried the most intuitive thing I could think of, in `./preprocessor.js`:

```
var ReactTools = require('react-tools');

module.exports =  function(src) {
    return ReactTools.transform(src);
}
```

I was hoping to get an error message that might lead me a step forward, but it looks like the compilers option was simply ignored.

Next attempt: Imitate the Coffeescript compiler, because it looks like that is working nicely for other people. Since the command for compiling Coffeescript is `mocha --compilers coffee:coffee-script/register`, I took a look at this crazy [require patcher thing](https://github.com/jashkenas/coffeescript/blob/master/src/register.coffee). It looks like a Mocha expects the compiler to patch `require`, but then the actual compilation step is obscured in the example by some process fork operation and by my lack of experience with Coffeescript in general. I am not understanding yet.

Third attempt: There is a 6to5 compiler with a register module written in a more familliar syntax, and it does not fork out to something else. It looks like the idea is to set `require.extensions['jsx'] = preprocess` where `preprocess` works like `function(m, filename) -> compiledJS`. I am so close! But for the life of me, I can't figure out what the `m` parameter is, and I feel like it is long past time to ask for help.

**TL;DR** How do I create a Mocha compiler from a `function compile(src) -> output`?

**Comment by neezer** [comment]
@d4goxn That was ***exactly*** what I was looking for. Thank you!

**Comment by kogai** [comment]
I seek exactly this! Thank you for @d4goxn 's report.
